File: Training/kompetitif/kenal/wildcard.py
#!/usr/bin/env python

patt = input()
n = int(input())
hasil = []

def ujung(depan = 0):
	global patt
	global n
	global hasil

	if depan:
		patt = patt[::-1]
	for i in range(n):
		if depan:
			a = input()[::-1]
		else:
			a = input()
		for j in range(len(patt)):
			if patt[j] == '*':
				if depan:
					print(a[::-1])
				else:
					print(a)
			else:
				if patt[j] != a[j]:
					break
	pass

def tengah():
	global patt
	global n
	global hasil


	for i in range(n):
		benar = 0
		a = input()
		for j in range(len(patt)):
			if patt[j] == '*':
				benar += 1
				break
			else:
				if patt[j] != a[j]:
					break

		patt = patt[::-1]
		a = a[::-1]
		for j in range(len(patt)):
			if patt[j] == '*':
				benar += 1
				if benar == 2:
					print(a[::-1])
			else:
				if patt[j] != a[j]:
					break

		patt = patt[::-1]
	pass

def all():
	global patt
	global n
	global hasil

	for i in range(n):
		a = input()
		print(a)
	pass

if patt[0] == '*':
	if patt[-1] == '*':
		continue
	else:
		ujung(depan = 1)
else:
	if patt[-1] == '*':
		ujung()
	else:
		tengah()
	
# for j in hasil:
	# print(j)

File: Training/kompetitif/kenal/wildcard.cpp
#include <bits/stdc++.h>

using namespace std;

string patt;
string input;
int n;

bool solve()
{
	if(patt == "*")
		return true;
	else
		if(patt[0] == '*')
		{
			string temp = "";
			string temp2 = "";

			for (int i = input.length(); i >= 0 ; --i)
				temp += input[i];

			for (int i = patt.length(); i >= 0; --i)
				temp2 += patt[i];

			for (int i = 0; i < temp2.length(); ++i)
				if(temp2[i] == '*')
					return true;
				else
					if(temp2[i] != temp[i])
						return false;
		}

		else if(patt[patt.length() - 1] == '*')
		{
			for (int i = 0; i < patt.length(); ++i)
				if(patt[i] == '*')
					return true;
				else
					if(patt[i] != input[i])
						return false;
		}
		else
		{
			int benar = 0;
			for (int i = 0; i < patt.length(); ++i)
				if(patt[i] == '*')
				{
					benar++;
					break;
				}
				else
					if(patt[i] != input[i])
						return false;

			string temp = "";
			string temp2 = "";

			for (int i = input.length(); i >= 0 ; --i)
				temp += input[i];

			for (int i = patt.length(); i >= 0; --i)
				temp2 += patt[i];

			for (int i = 0; i < temp2.length(); ++i)
				if(temp2[i] == '*')
				{
					benar++;
					if(benar == 2)
						return true;
				}
				else
					if(temp2[i] != temp[i])
						return false;
		}
}

int main()
{

	cin >> patt;
	cin >> n;

	for (int i = 0; i < n; ++i)
	{
		cin >> input;
		if (solve())
			cout << input << endl;
	}

	return 0;
}